internal/services/v1/experimental_test.go

```diff
@@ -102,3 +105,98 @@ func randomBatch(min, max int) func() int {
 		return rand.Intn(max-min) + min
 	}
 }
+
+func TestBulkExportRelationships(t *testing.T) {
+	conn, cleanup, _, _ := testserver.NewTestServer(require.New(t), 0, memdb.DisableGC, true, tf.StandardDatastoreWithSchema)
+	client := v1.NewExperimentalServiceClient(conn)
+	t.Cleanup(cleanup)
+
+	nsAndRels := []struct {
+		namespace string
+		relation  string
+	}{
+		{tf.DocumentNS.Name, "viewer"},
+		{tf.FolderNS.Name, "viewer"},
+		{tf.DocumentNS.Name, "owner"},
+		{tf.FolderNS.Name, "owner"},
+		{tf.DocumentNS.Name, "editor"},
+		{tf.FolderNS.Name, "editor"},
+	}
+
+	totalToWrite := uint64(1_000)
+	batch := make([]*v1.Relationship, totalToWrite)
+	for i := range batch {
+		nsAndRel := nsAndRels[i%len(nsAndRels)]
+		batch[i] = rel(
+			nsAndRel.namespace,
+			strconv.Itoa(i),
+			nsAndRel.relation,
+			tf.UserNS.Name,
+			strconv.Itoa(i),
+			"",
+		)
+	}
+
+	ctx := context.Background()
+	writer, err := client.BulkImportRelationships(ctx)
+	require.NoError(t, err)
+
+	require.NoError(t, writer.Send(&v1.BulkImportRelationshipsRequest{
+		Relationships: batch,
+	}))
+
+	resp, err := writer.CloseAndRecv()
+	require.NoError(t, err)
+	require.Equal(t, totalToWrite, resp.NumLoaded)
+
+	testCases := []struct {
+		batchSize      uint32
+		paginateEveryN int
+	}{
+		{1_000, math.MaxInt},
+		{10, math.MaxInt},
+		{1_000, 1},
+		{100, 5},
+		{97, 7},
+	}
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("%d-%d", tc.batchSize, tc.paginateEveryN), func(t *testing.T) {
+			require := require.New(t)
+
+			var totalRead uint64
+			var cursor *v1.Cursor
+
+			var done bool
+			for !done {
+				streamCtx, cancel := context.WithCancel(ctx)
+
+				stream, err := client.BulkExportRelationships(streamCtx, &v1.BulkExportRelationshipsRequest{
+					OptionalLimit:  tc.batchSize,
+					OptionalCursor: cursor,
+				})
+				require.NoError(err)
+
+				for i := 0; i < tc.paginateEveryN; i++ {
+					batch, err := stream.Recv()
+					if errors.Is(err, io.EOF) {
+						done = true
+						break
+					}
+
+					require.NoError(err)
+					require.LessOrEqual(uint32(len(batch.Relationships)), tc.batchSize)
+					require.NotNil(batch.AfterResultCursor)
+					require.NotEmpty(batch.AfterResultCursor.Token)
+
+					cursor = batch.AfterResultCursor
+					totalRead += uint64(len(batch.Relationships))
+				}
+
+				cancel()
+			}
+
+			require.Equal(totalToWrite, totalRead)
```

Comment:
can you also add a set to track the relationships returned and ensure we get back all of the expected rels?